Implement spell check correction feature for user input countries

nswarup14 opened the pull request:
Implements #5 

**Changes**
If the user input country is off by a single letter, then the game prompts the user with the following message _"Did you mean <correct_spelling> (y/n)"_. On pressing 'y' keyboard button, the flag of the country is displayed indicating the correct answer and on pressing any other letter, the usual incorrect message is displayed.

nswarup14 commented:
I wish to list out what _off by a character_ could mean, for easier understanding. 
- If the just one character is off
- If two characters are off, but the length of the country is 7 letters or above
- If three characters are off, but the length of the country is 14 or more

In all the three cases, the assumption is that the length of the input must be the same as the length of the country we are trying to input.

We can perhaps discuss over what other criteria can be added to the above list. (Similarity checks can be slightly complex as well, rather than a simple letter-by-letter scan).

Thanks

pro-panda commented:
Thanks, tested. Not reviewed

Your `off-by-X` algorithm supports only character replacements, but not additions and deletions.
See https://norvig.com/spell-correct.html as a neat example.

You can use the input string as a threshold for choosing X, without worrying about the actual length. This would also fit in nicely with the example by Peter Norvig.

pro-panda commented on Countries.py:
```diff
@@ -225,10 +232,27 @@ def run(self):
+                            ctry.text(l, answer_fix)
+                            self.ctry.message = "You almost got it right, Good Job!"
+                        else:
+                            self.ctry.message = "Sorry " + self.ctry.answer + " is not on my list"
```
There is a missing comma after `Sorry . . .`

pro-panda commented on ctry.py:
```diff
@@ -111,27 +111,48 @@ def do_letter(self, l):
             g.answers[ind] = ''
             self.redraw()
         g.answers[ind] = ans
+        # print(g.answers)
```
You forgot to remove a debug statement perhaps. You can use `git add -p` to add chunks of lines, rather than files together

pro-panda commented on ctry.py:
```diff
@@ -111,27 +111,48 @@ def do_letter(self, l):
+            for i in range(0, len(ans)):
+                if(ctry_name[i] != ans[i]):
+                    similar_count += 1
+            if(similar_count == 1 or (similar_count == 2 and len(ans)>7) or (similar_count == 3 and len(ans)>14)):
```
Follow PEP8 and flake standards. This line is > 79 chars

nswarup14 commented:
@pro-panda 
I have incorporated the changes you had suggested. I have also updated my `self.check_similar()` function in the `ctry.py`. I have used Peter Norvig idea of a spell corrector.
I have ignored _transposes_ as mentioned in the article, https://norvig.com/spell-correct.html. I feel we should stick to _replaces, additions and deletions_ only for determining how close the user input is to the country.

I have also made a couple of changes to the design and display of comments, as the activity is played. 
Thanks

pro-panda commented:
Please use PEP8 style coding. You may use `autopep8` for the same

pro-panda commented:
Tested.

I entered `Norwa`, and got `Did you mean Norway ? (y/n)`. I tried pressing Y from the screen button but it didn't work. Workaround was to press `Y` from the keyboard.

pro-panda commented on Countries.py:
```diff
@@ -225,10 +232,27 @@ def run(self):
+                            g.answers[ind] = ans
+                            self.ctry.flag(ans)
+                            ctry.text(l, answer_fix)
+                            self.ctry.message = "Good job, check the correct spelling!"
```
Change it to, `"Good Job, " + ans + " is the right answer."

nswarup14 commented:
@pro-panda 
Thanks.
Yes, right now the Y button from the screen is not enabled when the popup message `Did you mean <correct_spelling> (y/n)` appears. 
I wanted to discuss if we could add a suitable button the screen for the same purpose. Pressing the Y button may not be right way, because that Y is meant to enter only the names of the countries. 

The present workaround can continue be in place, as it is an intuitive approach.

pro-panda commented:
> The present workaround can continue be in place, as it is an intuitive approach.

This would imply that the activity can not be used on systems without keyboards.

> Pressing the Y button may not be right way, because that Y is meant to enter only the names of the countries.

To me, the alphabets as keys on the screen, not designed for any specific purpose. Let's use them for pressing `y` or `n`

nswarup14 commented:
> This would imply that the activity can not be used on systems without keyboards

No @pro-panda , I meant it as one of the approaches. On systems without keyboard option, the default option would be to use the mouse or touchscreen. On systems with keyboard support, the _y/n_ key press is just one the ways.

nswarup14 commented:
@pro-panda 
I have added support for the mouse click to serve as an option to respond to popups. The user can now achieve this by clicking on the `_y/n_` buttons on the screen with a mouse as well, along with the keyboard support.

pro-panda commented:
Please resolve conflicts before this can be merged

nswarup14 commented:
@pro-panda 
I have removed the conflicts, but I'm unable to remove the merge commit. Kindly guide me in the process.
Thanks

pro-panda commented:
> I have removed the conflicts, but I'm unable to remove the merge commit. Kindly guide me in the process.

I'm not sure. Here's what I do,
 - `git pull origin --rebase master`
 - then resolve conflicts by hand.

@quozl

quozl commented:
Thanks for asking.  Being `git`, there are many ways to resolve this.  There are practices you can follow to make merges and conflicts less likely.  Each developer learns these as they go, often depending on what else happens on the repository.

It is made more complex by some developers using a remote name of _origin_ to refer to their own repository, instead of the _sugarlabs_ repository.  My preference is that you clone the _sugarlabs_ repository (which makes it _origin_), and then `git remote add mine` for your own repository.  @pro-panda's advice above presumes that _origin_ is the _sugarlabs_ repository.  Check with `git remote -v`.  If it isn't the _sugarlabs_ repository, then you need to use a different remote name in the pull.

The triggering event for your conflicts was the merging of a12fd4fac082a65688539e124615bb1b31a3a369 into the _master_ branch.  That in turn was because you had this pull request open when https://github.com/sugarlabs/countries-activity/pull/11 was merged.  That in turn was because you were working simultaneously in separate branches, _double-click_ for 11, and _fix-feature_ for 9.

Your response was to `git pull` with merges enabled, or `git merge`.  This created the merge commit, which ties together the two ancestries.

Now, GitHub does _not_ show ancestry chains that `git` contains, but GitHub does preserve all these ancestry chains.  You can look at the ancestry chain using various tools.  `git log --all --graph` is the simplest to use.  Please do that now.

Had your response been `git pull --rebase`, the new commit a12fd4fac082a65688539e124615bb1b31a3a369 would have been placed _under_ your commits in the _fix-feature_ branch, and each of your commits renumbered.  If there were no merge conflicts that arise, this would be instantly successful.

However, your tree at the point before the merge (i.e. 4dd6d7d) edits the same files as a12fd4fac082a65688539e124615bb1b31a3a369 in a way that can't be trivially resolved by `git`, so you will have got conflicts; which I've reproduced for the purpose of demonstrating;

```log
$ git pull --rebase origin master
From https://github.com/sugarlabs/countries-activity
 * branch            master     -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: Implement spell check correction feature for user input countries
Using index info to reconstruct a base tree...
M       Countries.py
M       ctry.py
M       g.py
warning: 3 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging ctry.py
CONFLICT (content): Merge conflict in ctry.py
Auto-merging Countries.py
CONFLICT (content): Merge conflict in Countries.py
error: Failed to merge in the changes.
Patch failed at 0001 Implement spell check correction feature for user input countries
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

`git diff` shows the details of the conflict; they are extensive.  `git diff --stat` says in part;

```log
 Countries.py | 283 ++++++++++++[...]+++
 ctry.py      |  28 +++++++++++++
 2 files changed, 311 insertions(+)
```

@pro-panda's advise to resolve conflicts by hand would work, but it is a lengthy process, because (a) one of the files had line endings changed by a text editor or IDE, and (b) the changes are extensive.

Another way is to do it in reverse; start with a the _master_ branch and cherry-pick each of your new commits in the _fix-feature_ branch;

```
git checkout -b test master
git cherry-pick ...
```

I've tried that and the first commit conflicts.  You can't really know which method will work best without a good knowledge of the commits and exactly what they change.

Increase your skills of reading a diff, if you haven't done so already.  `git` is a storage system for diffs, and that's how it is best understood.

Also, as the two conflicts I've hit in reproducing your trouble were both noting very large changes due to line endings being changed, please try not to change line endings unless there is no other development taking place.  Had you carefully reviewed `git diff` before `git commit`, you would have seen the huge change and it should have given you pause.

Hope that helps.

nswarup14 commented:
@quozl @pro-panda 
Thank you for the explanation. I shall look into `git diff` more and understand it better.

> Also, as the two conflicts I've hit in reproducing your trouble were both noting very large changes due to line endings being changed, please try not to change line endings unless there is no other development taking place. Had you carefully reviewed git diff before git commit, you would have seen the huge change and it should have given you pause.

Yes, that was my fault. I did not know about how line endings are handled in git and shall be more careful in the future.

Thanks!
